pkg/ingester/ingester.go

```diff
@@ -1174,6 +1173,89 @@ func (i *Ingester) Push(ctx context.Context, req *cortexpb.WriteRequest) (*corte
+				} else {
+					h = cortexpb.HistogramProtoToHistogram(hp)
+				}
+
+				if ref != 0 {
+					if _, err = app.AppendHistogram(ref, copiedLabels, hp.TimestampMs, h, fh); err == nil {
+						succeededSamplesCount++
+						continue
+					}
+				} else {
+					// Copy the label set because both TSDB and the active series tracker may retain it.
+					copiedLabels = cortexpb.FromLabelAdaptersToLabelsWithCopy(ts.Labels)
+					if ref, err = app.AppendHistogram(0, copiedLabels, hp.TimestampMs, h, fh); err == nil {
+						succeededSamplesCount++
+						continue
+					}
+				}
+
+				failedSamplesCount++
+
+				// Check if the error is a soft error we can proceed on. If so, we keep track
+				// of it, so that we can return it back to the distributor, which will return a
+				// 400 error to the client. The client (Prometheus) will not retry on 400, and
+				// we actually ingested all samples which haven't failed.
+				switch cause := errors.Cause(err); {
+				case errors.Is(cause, storage.ErrOutOfBounds):
+					sampleOutOfBoundsCount++
+					updateFirstPartial(func() error { return wrappedTSDBIngestErr(err, model.Time(hp.TimestampMs), ts.Labels) })
+					continue
+
+				case errors.Is(cause, storage.ErrOutOfOrderSample):
+					sampleOutOfOrderCount++
+					updateFirstPartial(func() error { return wrappedTSDBIngestErr(err, model.Time(hp.TimestampMs), ts.Labels) })
+					continue
+
+				case errors.Is(cause, storage.ErrDuplicateSampleForTimestamp):
+					newValueForTimestampCount++
+					updateFirstPartial(func() error { return wrappedTSDBIngestErr(err, model.Time(hp.TimestampMs), ts.Labels) })
+					continue
+
+				case errors.Is(cause, storage.ErrTooOldSample):
+					sampleTooOldCount++
+					updateFirstPartial(func() error { return wrappedTSDBIngestErr(err, model.Time(hp.TimestampMs), ts.Labels) })
+					continue
+
+				case errors.Is(cause, errMaxSeriesPerUserLimitExceeded):
+					perUserSeriesLimitCount++
+					updateFirstPartial(func() error { return makeLimitError(perUserSeriesLimit, i.limiter.FormatError(userID, cause)) })
+					continue
+
+				case errors.Is(cause, errMaxSeriesPerMetricLimitExceeded):
+					perMetricSeriesLimitCount++
+					updateFirstPartial(func() error {
+						return makeMetricLimitError(perMetricSeriesLimit, copiedLabels, i.limiter.FormatError(userID, cause))
+					})
+					continue
+				case errors.As(cause, &errMaxSeriesPerLabelSetLimitExceeded{}):
+					updateFirstPartial(func() error {
+						return makeMetricLimitError(perLabelsetSeriesLimit, copiedLabels, i.limiter.FormatError(userID, cause))
+					})
+					continue
+				}
+
+				// The error looks an issue on our side, so we should rollback
+				if rollbackErr := app.Rollback(); rollbackErr != nil {
+					level.Warn(logutil.WithContext(ctx, i.logger)).Log("msg", "failed to rollback on error", "user", userID, "err", rollbackErr)
+				}
+				return nil, wrapWithUser(err, userID)
+			}
+		} else {
+			nativeHistogramCount += len(ts.Histograms)
+		}
```

Comment:
This seems duplicated? Can we extract to a new function?

Reply:
Done